File: Leetcode 3701-3800/3738/3738.java
// Leetcode 3738: Longest Non-Decreasing Subarray After Replacing at Most One Element
// https://leetcode.com/problems/longest-non-decreasing-subarray-after-replacing-at-most-one-element/
// Solved on 21st of November, 2025
class Solution {
    /**
     * Finds the length of the longest non-decreasing subarray after replacing at most one element.
     *
     * @param nums The input array of integers.
     * @return The length of the longest non-decreasing subarray.
     */
    public int longestSubarray(int[] nums) {
        int n = nums.length;
        if (n == 1) {
            return 1;
        }

        int[] right = new int[n];
        right[n - 1] = 1;
        for (int i = n - 2; i >= 0; i--) {
            if (nums[i] <= nums[i + 1]) {
                right[i] = right[i + 1] + 1;
            } else {
                right[i] = 1;
            }
        }

        int ans = 1;
        int left = 0;

        for (int i = 0; i < n; i++) {
            int prevLen = left;
            int nextLen = (i < n - 1) ? right[i + 1] : 0;

            ans = Math.max(ans, prevLen + 1);
            ans = Math.max(ans, 1 + nextLen);

            if (i > 0 && i < n - 1 && nums[i + 1] >= nums[i - 1]) {
                ans = Math.max(ans, prevLen + 1 + nextLen);
            }

            if (i > 0 && nums[i] >= nums[i - 1]) {
                left++;
            } else {
                left = 1;
            }
            ans = Math.max(ans, left);
        }

        return ans;
    }
}

File: Leetcode 3701-3800/3738/3738.py
# Leetcode 3738: Longest Non-Decreasing Subarray After Replacing at Most One Element
# https://leetcode.com/problems/longest-non-decreasing-subarray-after-replacing-at-most-one-element/
# Solved on 21st of November, 2025
class Solution:
    def longestSubarray(self, nums: list[int]) -> int:
        """
        Calculates the length of the longest non-decreasing subarray after replacing at most one element.

        Args:
            nums (list[int]): The input list of integers.
        Returns:
            int: The length of the longest non-decreasing subarray.
        """
        n = len(nums)
        if n == 1:
            return 1

        leftLengths = [1] * n
        rightLengths = [1] * n

        for i in range(1, n):
            if nums[i] >= nums[i - 1]:
                leftLengths[i] = leftLengths[i - 1] + 1

        for i in range(n - 2, -1, -1):
            if nums[i] <= nums[i + 1]:
                rightLengths[i] = rightLengths[i + 1] + 1

        maxLength = max(leftLengths)

        for i in range(n):
            leftPart = leftLengths[i - 1] if i > 0 else 0
            rightPart = rightLengths[i + 1] if i < n - 1 else 0

            currentLength = max(leftPart + 1, rightPart + 1)

            if i > 0 and i < n - 1 and nums[i + 1] >= nums[i - 1]:
                currentLength = max(currentLength, leftPart + 1 + rightPart)

            maxLength = max(maxLength, currentLength)

        return maxLength